File: 1 basic/cpp/1031. 查验身份证(15).cpp
#include<iostream>
#include<cstdio>
using namespace std;

/*
权重分配为：{7，9，10，5，8，4，2，1，6，3，7，9，10，5，8，4，2}
Z：0 1 2 3 4 5 6 7 8 9 10
M：1 0 X 9 8 7 6 5 4 3 2
*/
int W[]={7,9,10,5,8,4,2,1,6,3,7,9,10,5,8,4,2};
char M[]={'1','0','X','9','8','7','6','5','4','3','2'};

bool valid(char s[]){
	int sum=0;
	for(int i=0;i<17;++i){
		if(!isdigit(s[i])) return false;
		sum += (s[i]-'0') * W[i];
	}
	if(s[17] != M[sum%11]) return false;
	return true;
}


int main(){
	char id[19];
	int N,flag=0;
	scanf("%d",&N);
	while(N--){
		scanf("%s",id);
		if(!valid(id)){
			printf("%s---\n",id);
			flag=1;
		}
	}
	if(flag==0) printf("All passed");
	
	return 0;
}

File: 1 basic/py/1031. 查验身份证(15).py
d = {0:'1',1:'0',2:'X',3:'9',4:'8',5:'7',
     6:'6',7:'5',8:'4',9:'3',10:'2'}
W = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
N = int(input())
L = []
for i in range(N ):
    id = input()
    sum,flag= 0,True
    for j in range(17):
        if id[j].isdigit() : sum += W[j] * int(id[j])
        else:
            L.append(id)
            flag = False
            break
    if flag and d[sum%11] != id[17]: L.append(id)
if len(L )==0: print('All passed')
else:
    for id in L: print(id)
    